File: modbus_client.cpp
#include <modbus/modbus.h>
#include <iostream>
#include <unistd.h>

#define SERVER_IP "192.168.1.18"  // Modbus server IP address
#define SERVER_PORT 5020            // Modbus TCP port
#define MODBUS_ID 1                // Slave ID

int main() {
    // Create a Modbus TCP context
    modbus_t *mb;
    mb = modbus_new_tcp(SERVER_IP, SERVER_PORT);
    if (mb == NULL) {
        std::cerr << "Unable to create the Modbus connection" << std::endl;
        return -1;
    }

    modbus_set_slave(mb, MODBUS_ID);

    if (modbus_connect(mb) == -1) {
        std::cerr << "Connection failed: " << modbus_strerror(errno) << std::endl;
        modbus_free(mb);
        return -1;
    }

    // Read 1 holding register starting from address 0
    uint16_t tab_reg[6];  // Array to hold the register value
    int rc = modbus_read_registers(mb, 2, 1, tab_reg);  // Read 6 register starting from address 0
    if (rc == -1) {
        std::cerr << "Read failed: " << modbus_strerror(errno) << std::endl;
    } else {
        std::cout << "Register value: " <<rc<< std::endl;
    }

    // Close the connection
    modbus_close(mb);
    modbus_free(mb);

    return 0;
}

File: modbus_client.py
from pyModbusTCP.client import ModbusClient

MODBUS_SERVER_IP = "192.168.212.171"
MODBUS_SERVER_PORT = 5020

# Create a Modbus client instance
client = ModbusClient(host=MODBUS_SERVER_IP, port=MODBUS_SERVER_PORT)

# Open connection to the Modbus server
cn = client.open()

# Check if connection is successful
if cn:
    print("Connection to Modbus server established.")
    
    # Read 10 holding registers starting from address 0
    result = client.read_holding_registers(0, 10)
    coil_result = client.read_coils(0, 10)
    # Check if the result is a valid list (indicating success)
    if result:
        print(f"Register Values: {result}")
        print(f"coil Values: {coil_result}")
        client.write_single_register(0,100)
        client.write_single_coil(9,True)
    else:
        print("Error reading holding registers.")
    
    # Close the connection to the Modbus server
    client.close()

else:
    print("Failed to connect to Modbus server.")
